recirq/optimize/mpg.py

```diff
@@ -0,0 +1,390 @@
+        max_evaluations: The maximum number of function evaluations to allow
+            before termination.
+
+    Returns:
+        Scipy OptimizeResult
+    """
+
+    if known_values is not None:
+        known_xs, known_ys = known_values
+        known_xs = [np.copy(x) for x in known_xs]
+        known_ys = [np.copy(y) for y in known_ys]
+    else:
+        known_xs, known_ys = [], []
+
+    if max_evaluations is None:
+        max_evaluations = np.inf
+
+    n = len(x0)
+    log_sigma = np.ones(n) * log_sigma_init
+    sigma = np.exp(log_sigma)
+
+    # set up lr schedule and optimizer
+    lr_schedule1 = ExponentialSchedule(
+        learning_rate, decay_steps=decay_steps, decay_rate=decay_rate, staircase=True
+    )
+    lr_schedule2 = ExponentialSchedule(
+        learning_rate, decay_steps=decay_steps, decay_rate=decay_rate, staircase=True
+    )
+    optimizer1 = Adam(lr_schedule1)
+    optimizer2 = Adam(lr_schedule2)
+
+    _, f = wrap_function(f, args)
+    res = OptimizeResult()
+    current_x = np.copy(x0)
+    res.x_iters = []  # initializes as lists
+    res.xs_iters = []
+    res.ys_iters = []
+    res.func_vals = []
+    res.fun = 0
+    total_evals = 0
+    num_iter = 0
+    message = None
+
+    # stats
+    history_max = -np.inf
+
+    while num_iter < max_iterations:
+        # get samples from the current policy to evaluate
+        z = np.random.randn(batch_size, n)
+        new_xs = sigma * z + current_x
+
+        if total_evals + batch_size > max_evaluations:
+            message = "Reached maximum number of evaluations."
+            break
+
+        # Evaluate points
+        res.xs_iters.append(new_xs)
+        new_ys = [f(x) for x in new_xs]
+        res.ys_iters.append(new_ys)
+        total_evals += batch_size
+        known_xs.extend(new_xs)
+        known_ys.extend(new_ys)
+
+        # Save function value
+        if save_func_vals:
+            res.func_vals.append(f(current_x))
+            res.x_iters.append(np.copy(current_x))
+            res.fun = res.func_vals[-1]
+
+        # current sampling radius (maximal)
+        max_radius = 0
+        for x in new_xs:
+            if np.linalg.norm(x - current_x) > max_radius:
+                max_radius = np.linalg.norm(x - current_x)
+
+        reward = [-y for y in new_ys]
+
+        # warmup steps control whether to use the model to estimate the f
+        if num_iter >= warmup_steps:
+            # Determine points to use to build model
+            model_xs = []
+            model_ys = []
+            # safer way without the `SVD` not converging
+            try:
+                for x, y in zip(known_xs, known_ys):
+                    if np.linalg.norm(x - current_x) < radius_coeff * max_radius:
+                        model_xs.append(x)
+                        model_ys.append(y)
+                model = _get_quadratic_model(model_xs, model_ys, x)
+
+                # get samples (from model)
+                z = np.random.randn(batch_size_model, n)
```

Comment:
It's usually useful to have deterministic versions of stochastic functions. Consider accepting a numpy random state as an optional argument. see `cirq.parse_random_state` for a helpful utility function and https://cirq.readthedocs.io/en/stable/generated/cirq.RANDOM_STATE_OR_SEED_LIKE.html#cirq.RANDOM_STATE_OR_SEED_LIKE for more details